### ISSUE TITLE
URL mismatch with content

### ISSUE THREAD

**Issue by AaronMT** [mozilla-mobile/focus-android#377]
Nexus 6 (7.1.1) (4e3131aff8eb792be7d645c75ec5fea98db891f3)

1. Type `netflix` in the address-bar
2. Tap `Search for netflix`
3. Tap the address-bar
4. Type `net` and hit the arrow on the keyboard for the auto-complete to `netflix.com`

The URL becomes mismatched with the content. The URL should read netflix.com but it shows the prior search results.

![device-2017-03-30-114247](https://cloud.githubusercontent.com/assets/102331/24512978/0458fa30-153e-11e7-8065-1b9363ce21b0.png)

**Comment by pocmo** [comment]
@ahunt Regression from the spoof fix?

**Comment by ahunt** [comment]
I've reopened because there are still some issues, which my latest patches *should* hopefully fix.

One issue is that we don't always get a request for the main page URL, e.g. for mozilla.org (which redirects to mozilla.org/LANG), webkit won't notify us of a request to the redirected URL. Which means we can't safely update to the redirected URL, until loading ends (at which point we can check WebView's actual URL). This is different to what webview does on iOS.

I don't know why some devices aren't updating the URL at all, my latest patches should result in the correct URL being shown at the end of loading - but there might still be differences during loading. I've got a branch with debug logging that would help at least see what webkit callbacks we're getting, it would be good to get the output from that for failing websites:
https://github.com/mozilla-mobile/focus-android/tree/ahunt/debugurl

**Comment by pocmo** [comment]
I've built the 'debugurl' branch and this is the log for the following steps:
* Enter `moz`, autocompleted to `mozilla.org `, pressing enter
* http://mozilla.org/ appears in the URL bar and doesn't change
* When clicking into the URL bar the URL to edit is: https://www.mozilla.org/en-US/

```
                    art  I  Late-enabling -Xcheck:jni
                 System  W  ClassLoader referenced unknown path: /data/app/org.mozilla.focus.debug-2/lib/arm64
                    art  W  Before Android 4.1, method android.graphics.PorterDuffColorFilter android.support.graphics.drawable.VectorDrawableCompat.updateTint
                            Filter(android.graphics.PorterDuffColorFilter, android.content.res.ColorStateList, android.graphics.PorterDuff$Mode) would have inc
                            orrectly overridden the package-private method in android.graphics.drawable.Drawable
                 Adreno  I  QUALCOMM build                   : a5b4970, If5818605d9
                         I  Build Date                       : 10/12/16
                         I  OpenGL ES Shader Compiler Version: XE031.09.00.04
                         I  Local Branch                     : N24D
                         I  Remote Branch                    :
                         I  Remote Branch                    :
                         I  Reconstruct Branch               :
         OpenGLRenderer  I  Initialized EGL, version 1.4
                         D  Swap behavior 1
     ApplicationLoaders  D  ignored Vulkan layer search path /data/app/com.android.chrome-2/lib/arm64:/data/app/com.android.chrome-2/base.apk!/lib/arm64-v8a fo
                            r namespace 0x7f90a870f0
         WebViewFactory  I  Loading com.android.chrome version 56.0.2924.87 (code 292408752)
       cr_LibraryLoader  I  Time to load native libraries: 6 ms (timestamps 4239-4245)
                         I  Expected native library version number "56.0.2924.87", actual native library version number "56.0.2924.87"
                         I  Expected native library version number "56.0.2924.87", actual native library version number "56.0.2924.87"
               chromium  I  [INFO:library_loader_hooks.cc(163)] Chromium logging enabled: level = 0, default verbosity = 0
      cr_BrowserStartup  I  Initializing chromium process, singleProcess=true
               cr_media  W  Requires BLUETOOTH permission
                 cr_Ime  I  ImeThread is enabled.
IInputConnectionWrapper  W  finishComposingText on inactive InputConnection
                         W  finishComposingText on inactive InputConnection
              URLUPDATE  D  started=http://mozilla.org/ actualurl=http://mozilla.org/
                         D  shouldInterceptRequest=http://mozilla.org/
                 libEGL  E  validate_display:99 error 3008 (EGL_BAD_DISPLAY)
              URLUPDATE  D  started=https://www.mozilla.org/ actualurl=http://mozilla.org/
      VideoCapabilities  W  Unrecognized profile 2130706433 for video/avc
                         W  Unrecognized profile 2130706434 for video/avc
                         I  Unsupported profile 4 for video/mp4v-es
  NetworkSecurityConfig  D  No Network Security Config specified, using platform default
              URLUPDATE  D  started=https://www.mozilla.org/en-US/ actualurl=http://mozilla.org/
      cr_BindingManager  W  Cannot call determinedVisibility() - never saw a connection for the pid: 13075
              URLUPDATE  D  shouldInterceptRequest=https://www.mozilla.org/media/js/site-bundle.1688e984bc3d.js
                         D  shouldInterceptRequest=https://www.mozilla.org/media/css/pebbles-bundle.64e615be553d.css
                         D  shouldInterceptRequest=https://www.mozilla.org/media/css/home-bundle.d04604b5ae8d.css
                         D  shouldInterceptRequest=https://www.mozilla.org/media/js/gtm-snippet-bundle.bb4ed1f72aad.js
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/pebbles/moz-wordmark-dark-reverse.2cbc28bb9895.svg
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/l10n/en-US/firefox/android/btn-google-play-high-res.87d9720bbc8b.png
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/l10n/en-US/firefox/ios/btn-app-store.1cfd5dba4a92.svg
                         D  shouldInterceptRequest=https://www.mozilla.org/media/js/common-bundle.fef29294f735.js
                         D  shouldInterceptRequest=https://www.mozilla.org/media/js/stub-attribution-bundle.24a3d38f42a0.js
                         D  shouldInterceptRequest=https://www.mozilla.org/media/js/home-bundle.bf4814d0b8ec.js
                         D  shouldInterceptRequest=https://www.mozilla.org/media/js/task-completion-rate-survey-bundle.a54463c5f0db.js
                         D  shouldInterceptRequest=https://www.googletagmanager.com/gtm.js?id=GTM-MW3R8V&l=dataLayer
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/pebbles/icon-menu-dark.d12612893396.svg
                         D  shouldInterceptRequest=https://www.mozilla.org/media/fonts/opensans-regular.668362de763a.woff2
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/home/2016/intro/intro-learning-600.1ad8b9287179.jpg
                         D  shouldInterceptRequest=https://www.mozilla.org/media/fonts/opensans-bolditalic.1809656fcafc.woff2
                         D  shouldInterceptRequest=https://www.mozilla.org/media/fonts/opensans-bold.5cf854f3d1c0.woff2
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/home/2016/speaker-high-res.34dd5b0e1f05.png
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/home/2016/vr-high-res.31e068165622.png
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/firefox/horizon/stars.8398dac91f60.svg
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/firefox/template/logo-large-high-res.6b70bc06eb5c.png
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/firefox/horizon/foreground-high-res.1ab1938aae64.png
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/home/2016/icon-addons.013eb22dbe38.svg
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/home/2016/icon-careers.986c678e3af5.svg
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/home/2016/icon-help.6a0176e2ed1d.svg
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/newsletter/mozorg-newsletter.db230ae1e11b.svg
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/pebbles/moz-wordmark-light-reverse.cb1bdf6d1de6.svg
                         D  pagefinished=https://www.mozilla.org/en-US/ actualurl=https://www.mozilla.org/en-US/
                         D  shouldInterceptRequest=https://www.google-analytics.com/analytics.js
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/favicon.d4f1f46b91f4.ico
                         D  shouldInterceptRequest=https://www.mozilla.org/media/img/favicon/favicon-196x196.c80e6abe0767.png
                         D  shouldInterceptRequest=https://www.google-analytics.com/plugins/ua/linkid.js
                         D  shouldInterceptRequest=https://www.google-analytics.com/r/collect?v=1&_v=j49&aip=1&a=879274778&t=pageview&_s=1&dl=https%3A%2F%2Fwww
                            .mozilla.org%2Fen-US%2F&ul=en-us&de=UTF-8&dt=Internet%20for%20people%2C%20not%20profit%20%E2%80%94%20Mozilla&sd=32-bit&sr=412x732&v
                            p=412x659&je=0&_u=SCCAAEAjI~&jid=740983867&gjid=2133113685&cid=532038538.1490892815&tid=UA-36116321-1&_r=1&gtm=GTM-MW3R8V&cd40=(not
                            %20set)&cd41=(not%20set)&cd42=(not%20set)&cd43=(not%20set)&cd44=(not%20set)&cd51=false&cd53=false&cd54=false&cd55=true&cd56=Mozilla
                            %2F5.0%20(Linux%3B%20Android%207.1.1)%20AppleWebKit%2F537.36%20(KHTML%2C%20like%20Gecko)%20Version%2F4.0%20Focus%2F1.0%20Chrome%2F5
                            6.0.2924.87%20Mobile%20Safari%2F537.36&cd58=Homepage&cd59=52.0.2&z=447049137
```

**Comment by pocmo** [comment]
I can't reproduce the issue with a build from the `urlupdate`. The late update of the URL bar is a little bit unpleasant for slow loading websites - but it's better than no update. :)

I noticed that Fennec doesn't update the URL immediately too. At some point, while still loading, it switches to the new URL though.

**Comment by pocmo** [comment]
Did you look at other callbacks in WebViewClient / WebChromeClient too? I started with onPageStarted() / onPageFinished() because that's what I needed for progress - but maybe there are better ones that we can rely on for displaying the URL.
* For example onPageCommitVisible() sounds promising: `It is called at the earliest point at which it can be guaranteed that onDraw(Canvas) will no longer draw any content from previous navigations` - Argh, but it's only API 23+.
* Maybe the progress callbacks are helpful too? E.g. if we have loaded the next page partially then it's okay to switch the URL? Although this requires knowledge about what a specific progress value means.
